sound/soc/sof/topology.c

```diff
@@ -2379,15 +2379,17 @@ static int sof_link_unload(struct snd_soc_component *scomp,
 	struct snd_soc_dai_link *link =
 		container_of(dobj, struct snd_soc_dai_link, dobj);
 
-	struct snd_sof_dai *sof_dai = NULL;
+	struct snd_sof_dai *temp, *sof_dai = NULL;
 	int ret = 0;
 
-	list_for_each_entry(sof_dai, &sdev->dai_list, list) {
-		if (!sof_dai->name)
+	list_for_each_entry(temp, &sdev->dai_list, list) {
+		if (!temp->name)
```

Comment:
@bardliao does this really fix the problem? sof_dai will be assigned as the list is iterated so I dont see the use of a temp variable.

Reply:
@ranj063 the issue is in line 2395, we were using sof_dai as a matching flag at the same time, I think this is a good fix.

Reply:
@bardliao @keyonjie i see the problem. I think we can avoid the temp variable with a goto statement instead that points to the switch below.

Reply:
@ranj063 This PR is related to #343 . sof_link_unload() could be called with dai_link which is not in the dai_list. We won't find a dai in such case. And we use "if (!sof_dai)" in line 2395 to detect if we find a dai. But it won't be true even we can't find a dai.

Reply:
@ranj063 good idea. PR updated.